### PR TITLE
use the standard encode/decode scheme for Maybe and Tree

### LINKED ISSUE #107: Why is Maybe different?
The defacto standard set by `GSerialiseSum` is always a tag and an array. The handwritten types seem to keep to this.

Why is `Maybe` different? I'm having to special-case it in my generated C++ code (https://github.com/well-typed/binary-serialise-cbor/issues/89).

### PR DESCRIPTION
**Pull request by ghorn**
fixes #107, as requested

### CODE REVIEW

File: src/Data/Binary/Serialise/CBOR/Class.hs [hunk 1]
```diff
@@ -781,14 +781,16 @@ instance ( Serialise a, Serialise b, Serialise c, Serialise d, Serialise e
...
-                case n of
-                  0 -> return Nothing
-                  1 -> do !x <- decode
-                          return (Just x)
+                t  <- decodeWord
+                case (t, n) of
+                  (0, 1) -> return Nothing
+                  (0, _) -> fail "bad list length for Nothing"
+                  (1, 2) -> Just <$> decode
+                  (1, _) -> fail "bad list length for Just"
```
Comment (ghorn) [inline src/Data/Binary/Serialise/CBOR/Class.hs]: Is this how you want to handle the errors?

Comment (bgamari) [inline src/Data/Binary/Serialise/CBOR/Class.hs]: Looks reasonable.

**Comment by bgamari** [comment]
Sorry for the belated review!

I think this is about right. However, I wonder whether we should retain parse-compatibility with the old encoding. It would be relatively easy to do, although might make us susceptible to unexpectedly parsing what should be "invalid" CBOR. Given that we haven't actually released yet I'm rather on the fence about this.

**Comment by dcoutts** [comment]
So based on my comment from #107 I vote against this. We have to deal with some non-uniformity anyway and saving space on common types like Maybe is an advantage.

**Comment by bgamari** [comment]
Alright, I think the principle articulated in #107 is quite reasonable. Let's keep the encoding as-is.

**Comment by bgamari** [comment]
Regardless, thanks for your help, @ghorn!

**Comment by ghorn** [comment]
You're welcome! I agree with the decision fwiw

**Pull request closed without merging**